### ISSUE TITLE
[no-unused-vars] flagged though it is used inside an object

### ISSUE THREAD

**Issue by gforcada** [eslint/eslint#13327]
**Tell us about your environment**

* **ESLint Version:** 6.6.0 as well as 7.0.0
* **Node Version:** 12.16.3
* **npm Version:** 6.14.4

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "browser": true,
    "es6": true,
    "jquery": true
  },
  "extends": [
    "airbnb-base",
    "plugin:prettier/recommended"
  ],
  "parser": "babel-eslint",
  "parserOptions": {
    "sourceType": "module"
  },
  "plugins": ["prettier"],
  "rules": {
    "camelcase": "off",
  }
}
```
</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
  let placedAt = 0;
  let placedAt = myFunction({
    previous: placedAt
  });
```

```bash
eslint js/
```

**What did you expect to happen?**
eslint finishing without an error.

**What actually happened? Please include the actual, raw output from ESLint.**
eslint complains with ``error  'placedAt' is assigned a value but never used  no-unused-vars``.

While the `placedAt` variable is used inside the object given to `myFunction` function

**Are you willing to submit a pull request to fix this bug?**
I could try, though my time and JS-foo is not so good, but given some guidance I can give it a try.

**Comment by platinumazure** [comment]
Are you sure you want to redeclare `placedAt`?

Seems this would make more sense:

```js
  let placedAt = 0;
  placedAt = myFunction({
    previous: placedAt
  });
```

Also, no-unused-vars is expecting another use of `placedAt` after the reassignment. With the code you have provided, you could call `myFunction` without assigning the result, and this would both satisfy the rule and have no observable effect on behavior:

```js
  let placedAt = 0;
 myFunction({
    previous: placedAt
  });
```

If you do intend to reassign to `placedAt`, simply use it later in the code, and the rule will be satisfied.

**Comment by gforcada** [comment]
@platinumazure thanks for the quick feedback!

I see that I removed too much to simplify the example, the code looks like this:

```js
  let placedAt = 0;
  placedAt = myFunction({
    previous: placedAt
  });
  placedAt = myFunction({
    previous: placedAt
  });
  placedAt = myFunction({
    previous: placedAt
  });
```
So the `myFunction` returns a position, which is needed for the next call.

**Comment by platinumazure** [comment]
Hi @gforcada,

The logic is the same as before: On the last reassignment, you could just call the function for its side effects, without reassigning the variable.

```js
  let placedAt = 0;
  placedAt = myFunction({
    previous: placedAt
  });
  placedAt = myFunction({
    previous: placedAt
  });
  /* placedAt = */ myFunction({
    previous: placedAt
  });
```

Or you can add another usage:

```js
  let placedAt = 0;
  placedAt = myFunction({
    previous: placedAt
  });
  placedAt = myFunction({
    previous: placedAt
  });
  placedAt = myFunction({
    previous: placedAt
  });

  doSomethingWith(placedAt);
```

Basically, the last assignment is what is unused here. Either drop that reassignment, or use the variable again, and the rule will be satisfied.

**Comment by gforcada** [comment]
Oh I see know :facepalm: sorry for the noise and thanks for the (very quick :+1: ) explanations!! :bow:

**Comment by gforcada** [comment]
Final minor disclaimer that might be worthy: I did report it as `eslint` was complaining of `no-unused-vars` on the line where `placedAt` was being defined, not on the last usage that is indeed superfluous. Maybe that could be updated/improved (the error line where it is reported)?  :thinking:

**Comment by mdjermanovic** [comment]
> Final minor disclaimer that might be worthy: I did report it as `eslint` was complaining of `no-unused-vars` on the line where `placedAt` was being defined, not on the last usage that is indeed superfluous. Maybe that could be updated/improved (the error line where it is reported)?

There's an enhancement proposal for this: #13181
